File: src/benchscofi/LRSSL.py
#coding: utf-8

from stanscofi.models import BasicModel, create_scores
from stanscofi.preprocessing import CustomScaler

import numpy as np
from subprocess import call
from functools import reduce

import calendar
import time
current_GMT = time.gmtime()

class LRSSL(BasicModel):
    def __init__(self, params=None):
        params = params if (params is not None) else self.default_parameters()
        super(LRSSL, self).__init__(params)
        self.scalerS, self.scalerP = None, None
        self.sep_feature = params.get("sep_feature", "-")
        self.name = "LRSSL"
        self.model = None
        self.k, self.mu, self.lam, self.gam, self.maxiter, self.tol = [params[p] for p in ["k","mu","lam","gam","maxiter","tol"]]
        self.use_masked_dataset = True

    def default_parameters(self):
        params = {
            "decision_threshold": 2e-4, 
            "k": 10, "mu": 0.01, "lam": 0.01, "gam": 2, "tol": 1e-2, "maxiter": 8000, 
            "sep_feature": "-", 
        }
        return params

    def preprocessing(self, dataset, inf=2):
        if (self.scalerS is None):
            self.scalerS = CustomScaler(posinf=inf, neginf=-inf)
        S_ = self.scalerS.fit_transform(dataset.items.T.copy(), subset=None)
        if (self.scalerP is None):
            self.scalerP = CustomScaler(posinf=inf, neginf=-inf)
        P_ = self.scalerP.fit_transform(dataset.users.T.copy(), subset=None)
        if (all([self.sep_feature in str(f) for f in dataset.item_features])):
            types_feature = [str(f).split(self.sep_feature)[0] for f in dataset.item_features]
            X_lst = [S_[:,np.argwhere(np.array(types_feature)==tf)].T for tf in list(set(types_feature))]
        else:
            X_lst = [S_.T]
        X_p = P_ if (P_.shape[0]==P_.shape[1]) else np.corrcoef(P_)
        S_lst = [x.T.dot(x)/np.sqrt(x.sum(axis=0).dot(x.sum(axis=0).T)) for x in X_lst]
        S_lst = [s-np.diag(np.diag(s)) for s in S_lst]
        Y = dataset.ratings_mat.copy()
        ST = np.zeros((Y.shape[0], Y.shape[0]))
        for i in range(Y.shape[0]):
            for j in range(i+1,Y.shape[0]):
                s = X_p[Y[i,:]==1,:][:,Y[j,:]==1]
                if (s.shape[1]==0 or s.shape[0]==0):
                    continue
                ST[i,j] = np.max(s)
        ST = ST + ST.T
        S_lst.append(ST)
        X_lst = [x.T if (x.shape[0]==x.shape[1]) else np.corrcoef(x.T) for x in X_lst]
        return X_lst, S_lst, Y
        
    def fit(self, train_dataset):
        X_lst, S_lst, Y = self.preprocessing(train_dataset, inf=2)
        time_stamp = calendar.timegm(current_GMT)
        cmd = "mkdir -p %s/ && wget -qO - \'https://raw.githubusercontent.com/LiangXujun/LRSSL/a16a75c028393e7256e3630bc8b7900026061f99/LRSSL.R\' | sed -n \'/###/q;p\' > %s/LRSSL.R" % (time_stamp, time_stamp)
        call(cmd, shell=True)
        L_lst = []
        for s in S_lst:
            np.savetxt("%s/s.csv" % time_stamp,s)
            call("R -q -e 'source(\"%s/LRSSL.R\");S <- as.matrix(read.table(\"%s/s.csv\", sep=\" \", header=F));Sp <- get.knn.graph(S, %d);write.csv(S, \"%s/s.csv\", row.names=F)' 2>&1 >/dev/null" % (time_stamp, time_stamp, self.k, time_stamp), shell=True)
            L_lst.append(np.loadtxt("%s/s.csv" % time_stamp, skiprows=1, delimiter=","))
        L_lst = [np.diag(L.sum(axis=0))-L for L in L_lst]
        for i, x in enumerate(X_lst):
            np.savetxt("%s/x_%d.csv" % (time_stamp, i+1),x)
        for i, l in enumerate(L_lst):
            np.savetxt("%s/l_%d.csv" % (time_stamp, i+1),l)
        np.savetxt("%s/y.csv" % time_stamp, Y)
        assert all([s.shape[0]==Y.shape[0] and s.shape[1]==Y.shape[0] for s in S_lst])
        assert all([l.shape[0]==Y.shape[0] and l.shape[1]==Y.shape[0] for l in L_lst])
        assert all([x.shape[0]==Y.shape[0] and x.shape[1]==Y.shape[0] for x in X_lst])
        call("R -q -e 'source(\"%s/LRSSL.R\");ml <- %d;mx <- %d;X_lst <- lapply(1:mx, function(i) as.matrix(read.table(paste0(\"%s/x_\",i,\".csv\"), sep=\" \", header=F)));L_lst <- lapply(1:ml, function(i) as.matrix(read.table(paste0(\"%s/l_\",i,\".csv\"), sep=\" \", header=F)));Y <- as.matrix(read.table(\"%s/y.csv\", sep=\" \", header=F));train.res <- lrssl(X_lst, L_lst, Y, mx, ml, %f, %f, %f, %d, %f);for(i in 1:mx){write.csv(train.res$Gs[[i]], paste0(\"%s/G_\",i,\".csv\"), row.names=F)};write.csv(train.res$alpha, \"%s/alpha.csv\", row.names=F);write.csv(list(t=train.res$t,diff_G=train.res$diff.G), \"%s/vals.csv\", row.names=F);write.csv(train.res$F.mat, \"%s/F_mat.csv\", row.names=F)' | grep '\[1\]'" % (time_stamp, len(L_lst), len(X_lst), time_stamp, time_stamp, time_stamp, self.mu, self.lam, self.gam, self.maxiter, self.tol, time_stamp, time_stamp, time_stamp, time_stamp), shell=True)
        self.model = {
            "G": [np.loadtxt("%s/G_%d.csv" % (time_stamp, i+1), skiprows=1, delimiter=",") for i in range(len(X_lst))],
            "alpha": np.loadtxt("%s/alpha.csv" % time_stamp, skiprows=1, delimiter=","),
            "t": np.loadtxt("%s/vals.csv" % time_stamp, skiprows=1, delimiter=",")[0],
            "diff_G": np.loadtxt("%s/vals.csv" % time_stamp, skiprows=1, delimiter=",")[1],
            "F_mat": np.loadtxt("%s/F_mat.csv" % time_stamp, skiprows=1, delimiter=","),
        }
        call("rm -rf %s/" % time_stamp, shell=True)

    def model_predict(self, test_dataset):
        X_lst, _, _ = self.preprocessing(test_dataset, inf=2)
        Y_lst = [x.dot(self.model["G"][ix]) for ix, x in enumerate(X_lst)]
        Y_lst = [self.model["alpha"][iy]*(y/np.tile(y.sum(axis=1), (y.shape[1],1)).T) for iy, y in enumerate(Y_lst)]
        preds = np.ravel(reduce(sum, Y_lst))
        scores = create_scores(preds, test_dataset)
        return scores

File: src/benchscofi/implementations/LRSSL.R
#!/bin/R

## https://raw.githubusercontent.com/LiangXujun/LRSSL/a16a75c028393e7256e3630bc8b7900026061f99/LRSSL.R
get.knn <- function(v, k){
  ind <- order(v, decreasing = T)
  return(ind[1:k])
}

get.knn.graph <- function(S, k){
  n <- nrow(S)
  S.knn <- matrix(0, nrow = n, ncol = n)
  for(i in 1:n){
    ind <- get.knn(S[i,], k)
    S.knn[i,ind] <- 1
    S.knn[ind,i] <- 1
    #S.kkn[i,ind] <- S[i,ind]
    #S.knn[ind,i] <- S[i,ind]
  }
  return(S.knn)
}

lrssl <- function(Xs, Ls, Y, mx, ml, mu, lam, gam, max.iter, eps){
  n <- nrow(Y)
  c <- ncol(Y)
  Gs <- vector("list", mx)
  e1ds <- vector("list", mx)
  ds <- rep(0, mx)
  for(i in 1:mx){
    ds[i] <- nrow(Xs[[i]])
    Gs[[i]] <- matrix(runif(ds[i]*c), nrow = ds[i], ncol = c)
    e1ds[[i]] <- rep(1, ds[i])
  }
  alpha <- rep(1, ml)/ml
  
  check.step <- 20;
  Gs.old <- vector("list", mx)
  As <- vector("list", mx)
  As.pos <- vector("list", mx)
  As.neg <- vector("list", mx)
  Bs <- vector("list", mx)
  Bs.pos <- vector("list", mx)
  Bs.neg <- vector("list", mx)
  L <- matrix(0, nrow = n, ncol = n)
  for(i in 1:ml){
    L <- L + alpha[i]^gam*Ls[[i]]
  }
  
  t <- 0
  while(t < max.iter){
    t <- t + 1
    Q <- Y
    for(i in 1:mx){
      Gs.old[[i]] <- Gs[[i]]
      Q <- Q + mu*t(Xs[[i]])%*%Gs[[i]]
    }
    P <- solve(L + (1 + mx*mu)*diag(1, n, n))
    F.mat <- P%*%Q
    
    for(i in 1:mx){
      As[[i]] <- Xs[[i]]%*%(mu*diag(1,n,n)-mu^2*t(P))%*%t(Xs[[i]]) + lam*(e1ds[[i]]%*%t(e1ds[[i]]))
      As.pos[[i]] <- (As[[i]] + abs(As[[i]]))/2
      As.neg[[i]] <- (abs(As[[i]]) - As[[i]])/2
      Bs[[i]] <- mu*Xs[[i]]%*%P%*%Y
      for(j in 1:mx){
        if(i == j){
          next
        }else{
          Bs[[i]] <- Bs[[i]] + mu^2*Xs[[i]]%*%t(P)%*%t(Xs[[j]])%*%Gs[[j]]
        }
      }
      Bs.pos[[i]] <- (Bs[[i]] + abs(Bs[[i]]))/2
      Bs.neg[[i]] <- (abs(Bs[[i]]) - Bs[[i]])/2
    }
    for(i in 1:mx){
      Gs[[i]] <- Gs[[i]]*sqrt((Bs.pos[[i]] + As.neg[[i]]%*%Gs[[i]])/(Bs.neg[[i]] + As.pos[[i]]%*%Gs[[i]]))
    }
    
    for(i in 1:ml){
      alpha[i] <- (1/sum(diag(t(F.mat)%*%Ls[[i]]%*%F.mat)))^(1/(gam - 1))
    }
    alpha <- alpha/sum(alpha)
    
    L <- matrix(0, nrow = n, ncol = n)
    for(i in 1:ml){
      L <- L + alpha[i]^gam*Ls[[i]]
    }
    
    diff.G <- rep(0, mx)
    for(i in 1:mx){
      diff.G[i] <- norm(Gs[[i]] - Gs.old[[i]], "f")/norm(Gs.old[[i]], "f")
    }
    
    if(t%%check.step == 0){
      mesg <- sprintf("t = %i, diffG mean = %.4e", t, mean(diff.G))
      print(mesg)
    }
    if(mean(diff.G) < eps)
      break
  }
  return(list(Gs = Gs, F.mat = F.mat, alpha = alpha, diff.G = diff.G, t = t))
}

## Test

k <- 10
mu <- 0.01
lam <- 0.01
gam <- 2
tol <- 1e-2
maxiter <- 1000

## TRAIN

Xs <- as.matrix(read.table("../../Xs_df.csv", sep = " ", header = F))
Xp <- as.matrix(read.table("../../Xp_df.csv", sep = " ", header = F))
Y <- as.matrix(read.table("../../Y_df.csv", sep = " ", header = F))

X_lst <- list(Xs)

S_lst <- lapply(X_lst, function(X) t(X)%*%X/sqrt(colSums(X)%*%t(colSums(X))))
S_lst <- lapply(S_lst, function(S) S - diag(diag(S)))

ST <- matrix(0, nrow = nrow(Y), ncol = nrow(Y))
for(i in 1:(nrow(Y)-1)){
  for(j in (i+1):nrow(Y)){
    s <- Xp[Y[i,]==1,Y[j,]==1]
    ST[i,j] <- max(s)
  }
}
ST <- ST + t(ST)
S_lst[[length(S_lst)+1]] <- ST

S_knn_lst <- lapply(S_lst, function(S) get.knn.graph(S, k))
L_lst <- lapply(S_knn_lst, function(Sknn) diag(colSums(Sknn))-Sknn)

train.res <- lrssl(X_lst, L_lst, Y, length(X_lst), length(L_lst), mu, lam, gam, maxiter, tol)

## PREDICT

Xs <- as.matrix(read.table("../../Xs_df.csv", sep = " ", header = F))
Xp <- as.matrix(read.table("../../Xp_df.csv", sep = " ", header = F))
Y <- as.matrix(read.table("../../Y_df.csv", sep = " ", header = F))

X_lst <- list(Xs)

Y.pred_lst <- lapply(1:length(X_lst), function(i) X_lst[[i]]%*%train.res$Gs[[i]])
Y.pred_lst <- lapply(Y.pred_lst, function(Y) Y/rowSums(Y))
Y.pred_lst <- lapply(1:length(X_lst), function(i) train.res$alpha[i]*Y.pred_lst[[i]])
Y.pred <- Reduce("+", Y.pred_lst)